addons/binding/org.openhab.binding.allplay/src/main/java/org/openhab/binding/allplay/handler/AllPlayHandler.java

```diff
@@ -0,0 +1,478 @@
+
+    private final Logger logger = LoggerFactory.getLogger(AllPlayHandler.class);
+    private final AllPlay allPlay;
+    private Speaker speaker;
+    private VolumeRange volumeRange;
+
+    private final SpeakerReconnectService reconnectService;
+
+    private static final int VOLUME_STEP_SIZE = 5; // TODO: Make volume step change configurable
+    private static final int REWIND_SKIP_TIME_IN_MS = 10000; // TODO: Make configurable
+    private static final int FAST_FORWARD_SKIP_TIME_IN_MS = 10000; // TODO: Make configurable
+
+    public AllPlayHandler(Thing thing, AllPlay allPlay) {
+        super(thing);
+        this.allPlay = allPlay;
+        reconnectService = new SpeakerReconnectService(this);
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing AllPlay handler for speaker " + getDeviceId());
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Waiting for speaker to be discovered");
+        try {
+            allPlay.addSpeakerAnnouncedListener(this);
+            discoverSpeaker();
+        } catch (DiscoveryException e) {
+            logger.error("Uable to discover speakers", e);
+        }
+    }
+
+    public void discoverSpeaker() {
+        try {
+            logger.debug("Starting discovery for speaker " + getDeviceId());
+            allPlay.discoverSpeaker(getDeviceId());
+        } catch (DiscoveryException e) {
+            logger.error("Unable to discover speaker " + getDeviceId(), e);
+        }
+    }
+
+    @Override
+    public void onSpeakerAnnounced(Speaker speaker) {
+        logger.debug("Speaker announcement received for " + speaker + ". Own id is " + getDeviceId());
+        if (isHandledSpeaker(speaker)) {
+            this.speaker = speaker;
+            reconnectService.cancelReconnectionJob();
+            try {
+                connect();
+            } catch (AllPlayException e) {
+                logger.error(
+                        "Error while connecting to speaker" + speaker + ": " + e.getMessage() + ", scheduling retry");
+                logger.debug("Connection error", e);
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Error while communicating with speaker" + speaker + ": " + e.getMessage());
+                reconnectService.scheduleReconnectionJob(speaker);
+            }
+        }
+    }
+
+    private void connect() throws ConnectionException {
+        if (speaker != null) {
+            logger.info("Connecting to speaker " + speaker);
+            speaker.addSpeakerChangedListener(this);
+            speaker.addSpeakerConnectionListener(this);
+            speaker.connect();
+            logger.info("Connected to speaker " + speaker + " in thread " + Thread.currentThread().getName());
+            try {
+                updateStatus(ThingStatus.ONLINE);
+            } catch (Throwable t) {
```

Comment:
@kaikreuzer I'm having some issue here with the updateStatus, do you have an idea or know somebody, who could help?
In the initialize method, I am calling allPlay.discoverSpeaker(), which is asynchroneous and calling onSpeakerAnnounced() as soon as a speaker is discovered. This is also where I set the thing status to online. This works correctly if I restart openHAB or add a speaker for the first time. But if I remove a speaker thing and re-add it, the updateStatus fails with `java.lang.IllegalStateException: Could not update status, because callback is missing`.

Am I doing something wrong here? I added some debug and saw that the callback is actually set before the updatetStatus is called, however in another thread. But when I debug into the updateStatus method, the callback is in fact null.